File: foriOS/src/demo/DemoMediaCapture.m
/* Copyright (c) 2015 Good Technology Corporation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#import "DemoMediaCapture.h"
#import "DemoUtility.h"

#import <GD/GDFileManager.h>
#import <GD/GDAppServer.h>

#import "Backgrounder.h"
#import "BackgrounderForGoodDynamics.h"
#import "GdcRuntimeDispatcher.h"

#import "AVCamViewController.h"

@interface DemoMediaCapture()
@property (readonly)NSString *kRecordingsDirectory;
@property (readonly)NSString *kPhotosDirectory;

@property (nonatomic)DemoPanelItem *stopItem;
@property (nonatomic)NSArray *stopItemLocation;

@property (nonatomic)NSString *uploadURL;
@property (assign, nonatomic)BOOL observingConfiguration;
@end

@implementation DemoMediaCapture

static id<DemoUserInterface> demoMediaCaptureUserInterface = nil;

-(instancetype)init
{
    self = [super init];
    
    _gdcCommunication = [[GdcCommunication alloc] init];
    _kPhotosDirectory = [[DemoUtility documentsDirectory]
                         stringByAppendingPathComponent:@"photos"];
    _kRecordingsDirectory =[[DemoUtility documentsDirectory]
                            stringByAppendingPathComponent:@"recordings"];
    
    return self;
}

static void demoMediaCaptureLogger(NSString *message)
{
    [demoMediaCaptureUserInterface
     demoLogFormat:@"%@ %@", [DemoUtility simpleDate], message];
}

-(void)addStopItem
{
    self.stopItemLocation = [self appendToDivision:0 item:self.stopItem];
}
-(void)removeStopItem
{
    if (self.stopItemLocation == nil) return;

    NSMutableArray *division =
    self.demoPanelItemDivisions[[self.stopItemLocation[0] intValue]];
    
    [division removeObjectAtIndex:[self.stopItemLocation[1] intValue]];
    self.stopItemLocation = nil;
    
    [self demoPanelLoad:YES];
}

-(DemoPanelItem *)audioRecordingIn:(NSString *)directory
{
    DemoMediaCapture * __weak weakSelf = self;
    
    DemoPanelItemClick clickBlock = DEMOPANELITEMCLICK(demoPanelItem) {
        NSString *path = [DemoUtility numberedFileIn:directory
                                                stub:@"capture"
                                           extension:@"aiff"];
        if (path == nil) {
            [self.demoUserInterface demoLogFormat:
             @"Couldn't generate numbered file in \"%@\".\n", directory];
            return;
        }
        [weakSelf addStopItem];
        [BackgrounderForGoodDynamics
         gdStartRecording:path
         in:[Backgrounder sharedInstance] ];
    };
    
    return [DemoPanelItem
            demoPanelItemOfType:DemoPanelItemTypeCommandOn
            label:@"Record"
            clickBlock:clickBlock];
}

-(DemoPanelItem *)cameraStoringIn:(NSString *)directory
{
    DemoMediaCapture * __weak weakSelf = self;
//    NSString *path = nil;
    
    DemoPanelItemClick clickBlock = DEMOPANELITEMCLICK(demoPanelItem) {
        NSString *path = [DemoUtility numberedFileIn:directory
                                                stub:@"capture"
                                           extension:@"jpeg"];
        if (path == nil) {
            [self.demoUserInterface demoLogFormat:
             @"Couldn't generate numbered file in \"%@\".\n", directory];
        }
        else {
            // Set up the prepare block, which will tell the camera preview
            // which control needs to be refreshed after the picture has been
            // taken.
            self.demoPanelSegueBlock = DEMOPANELSEQUEBLOCK(segue) {
                AVCamViewController *avCamViewController =
                (AVCamViewController *)segue.destinationViewController;
                
                avCamViewController.demoMediaCapture = weakSelf;
                avCamViewController.capturePath = path;
            };
            
            [self.demoUserInterface segueToSecondViewFrom:self];
        }
    };
    
    return [DemoPanelItem
            demoPanelItemOfType:DemoPanelItemTypeCommandOn
            label:@"Camera"
            clickBlock:clickBlock];
}

-(void)observeConfiguration
{
    if (self.observingConfiguration) return;
    
    [[GdcRuntimeDispatcher sharedInstance]
     addObserverForApplicationConfiguration:
     GDC_CONFIGURATION_OBSERVER(configDictionary, configString) {
         NSArray *servers =
         [configDictionary objectForKey:GDAppConfigKeyServers];
         GDAppServer *gdAppServer = (GDAppServer *)servers[0];
         NSString *newUploadURL = [NSString stringWithFormat:@"http://%@:%@",
                                  gdAppServer.server, gdAppServer.port];
         if (self.uploadURL == nil ||
             ![self.uploadURL isEqualToString:newUploadURL] )
         {
             self.uploadURL = newUploadURL;
             [self.demoUserInterface
              demoLogFormat:@"Will upload to \"%@\"\n", self.uploadURL];
         }
         
     } andInvokeNow:YES];

    self.observingConfiguration = YES;
}

-(void)listMediaDirectory:(NSString *)path withPlayButtons:(BOOL)withPlay
{
    NSError *error = nil;
    GDFileManager *gdFileManager = [GDFileManager defaultManager];

    BOOL isDirectory = NO;
    if ([gdFileManager fileExistsAtPath:path isDirectory:&isDirectory]) {
        if (!isDirectory) {
            // Path exists but isn't a directory. Delete it.
            if (![gdFileManager removeItemAtPath:path error:&error]) {
                [self.demoUserInterface
                 demoLogFormat:@"Failed to delete non-directory \"%@\". %@.\n",
                 path, error];
            }
            // Next if will create a directory instead.
        }
        
        // Else: It was a directory, do nothing.
        
    }
    else {
        isDirectory = NO;
    }

    if (!isDirectory) {
        if ([gdFileManager createDirectoryAtPath:path
                     withIntermediateDirectories:YES
                                      attributes:nil
                                           error:&error])
        {
            [self.demoUserInterface
             demoLogFormat:@"Created directory \"%@\"\n", path];
        }
        else {
            [self.demoUserInterface
             demoLogFormat:@"Failed to  create directory \"%@\". %@.\n",
             path, error];
        }
        
    }
    
    error = nil;
    NSArray *listing = [[gdFileManager contentsOfDirectoryAtPath:path
                                                           error:&error]
                         sortedArrayUsingComparator:
                         ^NSComparisonResult(id obj1, id obj2) {
                             return [(NSString *)obj1 compare:(NSString *)obj2];
                         }];
    
    if (listing != nil) for (int index=0; index<listing.count; index++) {
        NSString *listingi = listing[index];
        NSString *pathi = [path stringByAppendingPathComponent:listingi];
        long division =
        [self addDivisionWithItem:[DemoPanelItem
                                   demoPanelItemLabel:listingi]];
        
        if (withPlay) {
            [self
             appendToDivision:division
             item:[DemoPanelItem
                   demoPanelItemOfType:DemoPanelItemTypeCommandOn
                   label:@"Play"
                   clickBlock:DEMOPANELITEMCLICK(demoPanelItem) {
                       [self addStopItem];
                       [BackgrounderForGoodDynamics
                        gdStartPlayback:pathi
                        in:[Backgrounder sharedInstance]];
                       
                   }]];
        }
        
        [self
         appendToDivision:division
         item:[DemoPanelItem
               demoPanelItemOfType:DemoPanelItemTypeCommandOn
               label:@"Upload"
               clickBlock:DEMOPANELITEMCLICK(demoPanelItem) {
                   [self.gdcCommunication upload:pathi
                                              to:self.uploadURL];
               }]
         ];
        
        [self
         appendToDivision:division
         item:[DemoPanelItem demoPanelItemLabel:[DemoUtility
                                                 statFile:pathi]]
         ];

        [self
         appendToDivision:division
         item:[DemoPanelItem
               demoPanelItemOfType:DemoPanelItemTypeCommandOn
               label:@"Delete"
               clickBlock:DEMOPANELITEMCLICK(demoPanelItem) {
                   NSError *blockError = nil;
                   if (![gdFileManager removeItemAtPath:pathi
                                                  error:&blockError])
                   {
                       [self.demoUserInterface
                        demoLogFormat:@"Failed to delete \"%@\". %@.\n",
                        pathi, blockError];
                   }
                   [self demoPanelLoad:YES];
               }]
         ];
        
    }
    
    if (error != nil) {
        [self.demoUserInterface
         demoLogFormat:@"Failed to list directory \"%@\". %@.\n",
         path, error];
    }
}

-(void)demoPanelLoad:(BOOL)andRefresh
{
    DemoMediaCapture * __weak weakSelf = self;
    
    demoMediaCaptureUserInterface = [self demoUserInterface];
    [[Backgrounder sharedInstance] setLogger:demoMediaCaptureLogger];

    [weakSelf observeConfiguration];
    
    self.gdcCommunication.logger = GDC_COMMUNICATION_LOGGER(message) {
        [weakSelf.demoUserInterface
         demoLogFormat:@"%@ %@", [DemoUtility simpleDate], message];
    };

    if (self.stopItem == nil) {
        self.stopItem = [DemoPanelItem
                         demoPanelItemOfType:DemoPanelItemTypeCommandBack
                         label:@"Stop"
                         clickBlock:DEMOPANELITEMCLICK(demoPanelItem){
                             [[Backgrounder sharedInstance] stop];
                             [weakSelf removeStopItem];
                         }];
    }
    
    [self deleteAllDivisions];
    
    [self addDivisionWithArray:
     @[ [DemoPanelItem demoPanelItemLabel:@"Media Capture"],
        [weakSelf audioRecordingIn:self.kRecordingsDirectory],
        [weakSelf cameraStoringIn:self.kPhotosDirectory] ]
     ];
    
    [weakSelf listMediaDirectory:self.kPhotosDirectory withPlayButtons:NO];
    [weakSelf listMediaDirectory:self.kRecordingsDirectory withPlayButtons:YES];
    
    if (andRefresh) {
        [self.demoUserInterface refresh];
    }
    
}

@end

File: forAndroid/contributorLibrary/src/main/java/com/good/example/contributor/jhawkins/demo/DemoMediaCapture.java
/* Copyright (c) 2015 Good Technology Corporation
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package com.good.example.contributor.jhawkins.demo;

import com.good.example.contributor.jhawkins.demoframework.Panel;
import com.good.example.contributor.jhawkins.demoframework.PanelItem;
import com.good.example.contributor.jhawkins.demoframework.PanelItemType;
import com.good.example.contributor.jhawkins.gdccommunication.GdcHttp;
import com.good.example.contributor.jhawkins.gdccommunication.ProgressListener;
import com.good.example.contributor.jhawkins.gdcmedia.CameraPreview;
import com.good.example.contributor.jhawkins.gdcmedia.Playback;
import com.good.example.contributor.jhawkins.gdcmedia.Recording;
import com.good.example.contributor.jhawkins.gdcmedia.Utility;
import com.good.example.contributor.jhawkins.gdruntime.RuntimeDispatcher;
import com.good.gd.GDAndroid;
import com.good.gd.GDAppServer;
import com.good.gd.file.File;

import java.io.IOException;
import java.util.List;
import java.util.Map;

public class DemoMediaCapture extends Panel {

    private PanelItem stopItem = null;
    private int[] stopItemLocation = null;

    private Recording recording = null;
    private Playback playback = null;
    Utility.Logger mediaLogger = null;

    private String uploadURL = null;
    private RuntimeDispatcher.ConfigurationHandler configurationHandler = null;

    private class PanelItemStop extends PanelItem {
        public PanelItemStop () {
            super(PanelItemType.COMMANDBACK, "Stop");
        }

        public void onClick() {
            this.panel.userInterface().demoLog(
                    this.getClass().getSimpleName() + ".onClick(" + userData + ")\n");

            if (recording != null) recording.stop();
            if (playback != null) playback.stop();

            removeStopItem();

            // Todo: Replace this with a completion handler in the Recording, or Playback.
            demoPanelLoad(false);
        }
    }

    public void addStopItem() {
        if (stopItemLocation != null) return;

        if (stopItem == null) {
            stopItem = new PanelItemStop();
        }
        stopItemLocation = appendToDivision(0, stopItem);
    }

    public PanelItem removeStopItem() {
        if (stopItemLocation == null) return null;
        PanelItem panelItem = removeItemAt(stopItemLocation[0], stopItemLocation[1]);
        stopItemLocation = null;
        return panelItem;
    }

    private Utility.Logger getMediaLogger() {
        if (mediaLogger == null)
            mediaLogger = new Utility.Logger() {
                @Override
                public void logMessage(String message) {
                    DemoMediaCapture.this.userInterface().demoLog(message);
                }
            };
        return mediaLogger;
    }

    private class PanelItemRecord extends PanelItem {
        public PanelItemRecord (String userData) {
            super(PanelItemType.COMMANDON, "Record", userData);
        }

        public void onClick() {
            this.panel.userInterface().demoLog(
                    this.getClass().getSimpleName() + ".onClick(" + userData + ")\n");

            File directory = new File((String)userData);
            final File path = com.good.example.contributor.jhawkins.demo.Utility.
                    numberedFileIn(directory, "capture", "pcm");
            if (path == null) {
                userInterface().demoLog(
                        "Couldn't generate numbered file in \"" + directory + "\".\n");
                return;
            }

            recording = new Recording();
            recording.logger = getMediaLogger();
            String recordingError = recording.startOrError(path, getActivity());
            if (recordingError == null) {
                addStopItem();
            }
            else {
                this.panel.userInterface().demoLog(
                        this.getClass().getSimpleName() + " failed to start recording. " +
                        recordingError+ "\n");
            }
        }
    }

    private class PanelItemPlay extends PanelItem {
        public PanelItemPlay (String userData) {
            super(PanelItemType.COMMANDON, "Play", userData);
        }

        public void onClick() {
            this.panel.userInterface().demoLog(
                    this.getClass().getSimpleName() + ".onClick(" + userData + ")\n");

            playback = new Playback();
            playback.logger = getMediaLogger();
            String playbackError = playback.startOrError((String)userData, getActivity());
            if (playbackError == null) {
                addStopItem();
            }
            else {
                this.panel.userInterface().demoLog(
                        this.getClass().getSimpleName() + " failed to start playback. " +
                                playbackError+ "\n");
            }
        }

    }

    private class PanelItemCamera extends PanelItem {
        public PanelItemCamera (String userData) {
            super(PanelItemType.COMMANDON, "Camera", userData);
        }

        CameraPreview cameraPreview = null;

        @Override
        public void onClick() {
            userInterface().demoLog(
                    this.getClass().getSimpleName() + ".onClick(" + userData + ")\n");

            File directory = new File((String)userData);
            final File path = com.good.example.contributor.jhawkins.demo.Utility.
                    numberedFileIn(directory, "capture", "jpeg");
            if (path == null) {
                userInterface().demoLog(
                        "Couldn't generate numbered file in \"" + directory + "\".\n");
                return;
            }

            userInterface().activity().runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    cameraPreview = (CameraPreview)(userInterface().secondView());
                    cameraPreview.logger = getMediaLogger();

                    cameraPreview.runnableOnCapture = new Runnable() {
                        @Override
                        public void run() {
                            panel.demoPanelLoad(true);
                        }
                    };

                    cameraPreview.setCamera(path, userInterface().activity());
                }
            });

        }
    }

    private class PanelItemUpload extends PanelItem {
        public PanelItemUpload (File userData) {
            super(PanelItemType.COMMANDON, "Upload", userData);
        }

        private ProgressListener progressListener = null;
        private Boolean needNL = false;

        @Override
        public void onClick() {
            if (uploadURL == null) {
                userInterface().demoLog("No upload URL.\n");
                return;
            }

            File path = (File)userData;

            progressListener = new ProgressListener() {
                @Override
                public long getProgressStep() { return 10L * 0x8000L; }

                @Override
                public void started() {
                    userInterface().demoLog("Starting upload");
                    needNL = true;
                }

                @Override
                public void transferred(long transferredBytes) {
                    userInterface().demoLog(" " + transferredBytes);
                    needNL = true;
                }
            };

            try {
                String status = GdcHttp.upload(uploadURL, path, progressListener);
                userInterface().demoLog( (needNL ? "\n" : "") +
                        "Upload finished: " + status + ".\n");
            } catch (IOException exception) {
                userInterface().demoLog( (needNL ? "\n" : "") +
                        "Failed to upload \"" + path + "\". " + exception.toString());
            }
            needNL = false;
        }
    } // PanelItemUpload class.

    private class PanelItemDelete extends PanelItem {
        public PanelItemDelete(File userData) {
            super(PanelItemType.COMMANDON, "Delete", userData);
        }

        @Override
        public void onClick() {
            File path = (File)userData;

            if (!path.delete()) userInterface().demoLog(
                    "Failed to delete \"" + path.toString() + "\".\n");
            panel.demoPanelLoad(true);
        }
    }

    private void listMediaDirectory(String path, Boolean playButtons)
    {
        File directory = new File(path);

        if (!directory.exists()) {
            // Todo: Check whether there is a file already and delete it if there is.
            Boolean mkdirsOK = directory.mkdirs();
            if (mkdirsOK) {
                userInterface().demoLog("Created directory \"" + path + "\".\n");
            }
            else {
                userInterface().demoLog("Failed to create directory \"" + path + "\".\n");
            }
        }

        String[] listing = directory.list();
        for (String listingi : listing) {
            File file = new File(path, listingi);
            int division = addDivision( new PanelItem(file.getName()) );
            if (playButtons) {
                appendToDivision(division, new PanelItemPlay(file.toString()));
            }
            appendToDivision(division, new PanelItemUpload(file));
            appendToDivision(division, new PanelItem(Long.toString(file.length())));
            appendToDivision(division, new PanelItemDelete(file));
        }
    }

    private void observeConfiguration()
    {
        if (configurationHandler == null) {
            configurationHandler = new RuntimeDispatcher.ConfigurationHandler() {
                @Override
                public void onReceiveMessage(Map<String, Object> map, String string) {
                    List<GDAppServer> servers =
                            (List<GDAppServer>) map.get(GDAndroid.GDAppConfigKeyServers);
                    if (servers.size() <= 0) {
                        userInterface().demoLog("No server configured for upload.\n");
                        return;
                    }

                    String newURL = "http://" + servers.get(0).server + ":" + servers.get(0).port;
                    if (uploadURL == null || !uploadURL.equalsIgnoreCase(newURL)) {
                        uploadURL = newURL;
                        userInterface().demoLog("Will upload to \"" + uploadURL + "\"\n");
                    }
                }
            };
            RuntimeDispatcher.getInstance().addApplicationConfigurationHandler(
                    configurationHandler, true);
        }
    }

    @Override
    public void demoPanelLoad(Boolean andRefresh)
    {
        observeConfiguration();
        deleteAllDivisions();

        addDivision(
                new PanelItem("Media Capture"),
                new PanelItemRecord("recordings"),
                new PanelItemCamera("photos")
        );

        listMediaDirectory("photos", false);
        listMediaDirectory("recordings", true);

        if (andRefresh) {
            this.userInterface.refresh();
        }
    }

}
